### PR TITLE
Improving Scoping & Components

### PR DESCRIPTION
**Pull request by mikelikespie**
* Simplified DSL (configure function)
 * Binders now have scopes. One can only scope things for the binder
   that is scoping it
 * Scope has to be passed along to modules from component
 * Language can enforce scoping now! this means validations can be
   removed
 * Removed "Scoped" protocol. This created ambiguity

### CODE REVIEW

File: CleanseTests/ScopeErrorTests.swift [hunk 1]
```diff
@@ -20,39 +20,14 @@ struct Scope2 : Scope {
 
 class ScopeErrorTests: XCTestCase {
 
-    struct ComponentWithInvalidScope1 : RootComponent {
```
Comment (mikelikespie) [inline CleanseTests/ScopeErrorTests.swift]: These are all removed because it isn't possible to get into this state anymore

**Comment by mikelikespie** [comment]
Not quite sure if I like the name `scoped().` It kinda doesn't imply what it does. Maybe "shared"? Rx uses sharing to mean similar concept.

**Comment by mikelikespie** [comment]
ok renamed `scoped()` -> `sharedInScope()` cuz that's essentially what it does

File: Cleanse/BindingBuilder.swift [hunk 2]
```diff
@@ -70,37 +70,30 @@ extension Provider : _StandardProvider, _AnyStandardProvider {
...
 }
 
-extension BindingBuilder where MaybeScope == _Unscoped {
-    
-    public func asSingleton() -> ScopedBindingDecorator<Self, Singleton> {
-        return self.with()
+extension BindingBuilder where MaybeScope == Unscoped {
+    @available(*, unavailable, renamed: "sharedInScope()")
+    public func asSingleton() -> ScopedBindingDecorator<Self, Unscoped> {
+        preconditionFailure()
```
Comment (holmes) [inline Cleanse/BindingBuilder.swift]: Add a useful message?

File: Cleanse/BindingBuilder.swift [hunk 3]
```diff
@@ -70,37 +70,30 @@ extension Provider : _StandardProvider, _AnyStandardProvider {
...
     }
 
-    
+
+    @available(*, unavailable, renamed: "sharedInScope()")
     public func scoped<S: Scope>(`in` scope: S.Type) -> ScopedBindingDecorator<Self, S> {
-        return self.with()
+        preconditionFailure()
```
Comment (holmes) [inline Cleanse/BindingBuilder.swift]: useful message?

File: Cleanse/ProviderProvider.swift [hunk 4]
```diff
@@ -12,7 +12,7 @@ import Foundation
 /**
  Portion of the binder protocol used to request future providers. This is only used internally and will probably go away
  */
-public protocol ProviderProvider : class {
+public protocol ProviderProvider {
```
Comment (holmes) [inline Cleanse/ProviderProvider.swift]: Yay for removing `class`

File: Cleanse/BindingBuilder.swift (hunk 3 above)
Comment (mikelikespie) [inline Cleanse/BindingBuilder.swift]: it won't compile, so no error mesage required. Just needed something to make it not expect a return value

File: Cleanse/BindingBuilder.swift (hunk 2 above)
Comment (mikelikespie) [inline Cleanse/BindingBuilder.swift]: it won't compile, so no error mesage required. Just needed something to make it not expect a return value

**Pull request merged**